cpp/src/arrow/flight/sql/driver.cc

```diff
@@ -0,0 +1,1877 @@
+        next_endpoint_(0) {}
+
+  [[nodiscard]] std::shared_ptr<Schema> schema() const override { return schema_; }
+
+  Status ReadNext(std::shared_ptr<RecordBatch>* batch) override {
+    FlightStreamChunk chunk;
+    while (current_stream_ && !chunk.data) {
+      ARROW_ASSIGN_OR_RAISE(chunk, current_stream_->Next());
+      if (chunk.data) {
+        *batch = chunk.data;
+        break;
+      }
+      if (!chunk.data && !chunk.app_metadata) {
+        RETURN_NOT_OK(NextStream());
+      }
+    }
+    if (!current_stream_) *batch = nullptr;
+    return Status::OK();
+  }
+
+  Status Close() override {
+    if (current_stream_) {
+      current_stream_->Cancel();
+    }
+    return Status::OK();
+  }
+
+  AdbcStatusCode Init(struct AdbcError* error) {
+    ADBC_ARROW_RETURN_NOT_OK(IO, error, NextStream());
+    if (!schema_) {
+      // Empty result set - fall back on schema in FlightInfo
+      ipc::DictionaryMemo memo;
+      ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error, info_->GetSchema(&memo).Value(&schema_));
+    }
+    return ADBC_STATUS_OK;
+  }
+
+  /// \brief Export to an ArrowArrayStream
+  static AdbcStatusCode Export(FlightSqlClient* client, FlightCallOptions call_options,
+                               std::unique_ptr<FlightInfo> info,
+                               struct ArrowArrayStream* stream, struct AdbcError* error) {
+    auto reader = std::make_shared<FlightInfoReader>(client, std::move(call_options),
+                                                     std::move(info));
+    ADBC_RETURN_NOT_OK(reader->Init(error));
+    ADBC_ARROW_RETURN_NOT_OK(INTERNAL, error,
+                             ExportRecordBatchReader(std::move(reader), stream));
+    return ADBC_STATUS_OK;
+  }
+
+ private:
+  Status NextStream() {
+    if (next_endpoint_ >= info_->endpoints().size()) {
+      current_stream_ = nullptr;
+      return Status::OK();
+    }
+    const FlightEndpoint& endpoint = info_->endpoints()[next_endpoint_];
+
+    if (endpoint.locations.empty()) {
+      ARROW_ASSIGN_OR_RAISE(current_stream_,
+                            client_->DoGet(call_options_, endpoint.ticket));
+    } else {
+      // TODO(lidavidm): this should come from a connection pool
+      std::string failures;
+      for (const Location& location : endpoint.locations) {
+        auto status =
+            FlightClient::Connect(location, DefaultClientOptions()).Value(&data_client_);
+        if (!status.ok()) {
+          if (!failures.empty()) {
+            failures += "; ";
+          }
+          failures += location.ToString();
+          failures += ": ";
+          failures += status.ToString();
+          data_client_.reset();
+          continue;
+        }
+        break;
+      }
+
+      if (!data_client_) {
+        return Status::IOError("Failed to connect to all endpoints: ", failures);
+      }
+      ARROW_ASSIGN_OR_RAISE(current_stream_,
+                            data_client_->DoGet(call_options_, endpoint.ticket));
```

Comment:
We can't reset it here since we still need to use the stream (I guess the stream probably contains a strong reference to the client internally).

I'll add the fallback for DoGet as well, good point